---
layout: project
title: SRTsim/
project: SRTsim
---




Package Website: [SRTsim Tutorial](https://xzhoulab.github.io/SRTsim/)

# Overview
------
Spatially resolved transcriptomics (SRT)-specific computational methods are often developed, tested, validated, and evaluated in silico using simulated data. Unfortunately, existing simulated SRT data are often poorly documented, hard to reproduce, or unrealistic. Single-cell simulators are not directly applicable for SRT simulation as they cannot incorporate spatial information. We present SRTsim, an SRT-specific simulator for scalable, reproducible, and realistic SRT simulations. SRTsim not only maintains various expression characteristics of SRT data but also preserves spatial patterns. We illustrate the benefits of SRTsim in benchmarking methods for spatial clustering, spatial expression pattern detection, and cell-cell communication identification.

<br />
SRTsim is an independent, reproducible, and flexible SRT simulation framework that can be used to facilitate the development of SRT analytical methods for a wide variety of SRT-specific analyses. It utilizes spatial localization information to simulate SRT expression count data in a reproducible and scalable fashion. 
Two major simulation schemes are implemented in SRTsim: [reference-based](https://xzhoulab.github.io/SRTsim/02_Reference_Based_Example/) and [reference-free](https://xzhoulab.github.io/SRTsim/03_Reference_Free_Example/).

# Paper
------
**Paper:** SRTsim: spatial pattern preserving simulations for spatially resolved transcriptomics, Genome Biology, 2023
[https://link.springer.com/article/10.1186/s13059-023-02879-z](https://link.springer.com/article/10.1186/s13059-023-02879-z)

<img src="/assets/images/papers/SRTsim.png" alt="photo" width="500"/>


# Tutorial of SRTsim
------
[SRTsim Tutorial](https://xzhoulab.github.io/SRTsim/)



# Analysis Codes
------
https://xzhoulab.github.io/SRTsim/04_Reproduce/
