---
layout: paper
title: "A Review of Statistical Methods for Identifying Trait-Relevant Tissues and Cell Types"
image: /assets/images/papers/review1.jpg
authors: Huanhuan Zhu*, Lulu Shang*, Xiang Zhou
year: 2021
shortref: Zhu et al. (2021) 
journal: Frontiers in Genetics
type: review
doi: 
---

# Abstract

**Overview**
Genome-wide association studies (GWASs) have identified and replicated many genetic variants that are associated with diseases and disease-related complex traits. However, the biological mechanisms underlying these identified associations remain largely elusive. Exploring the biological mechanisms underlying these associations requires identifying trait-relevant tissues and cell types, as genetic variants likely influence complex traits in a tissue- and cell type-specific manner. Recently, several statistical methods have been developed to integrate genomic data with GWASs for identifying trait-relevant tissues and cell types. These methods often rely on different genomic information and use different statistical models for trait-tissue relevance inference.

**Results**
Here, we present a comprehensive technical review to summarize ten existing methods for trait-tissue relevance inference. These methods make use of different genomic information that include functional annotation information, expression quantitative trait loci information, genetically regulated gene expression information, as well as gene co-expression network information. These methods also use different statistical models that range from linear mixed models to covariance network models. 

**Conclusion**
We hope that this review can serve as a useful reference both for methodologists who develop methods and for applied analysts who apply these methods for identifying trait relevant tissues and cell types.

<br />

<div class="middle">
    <img src="/assets/images/papers/review1.jpg" alt="photo" width="500"/>
</div>

<br />

**Paper:**
<br />
[https://doi.org/10.3389/fgene.2020.587887](https://doi.org/10.3389/fgene.2020.587887)
<br />


