---
layout: paper
title: "SRTsim: spatial pattern preserving simulations for spatially resolved transcriptomics"
image: /assets/images/papers/SRTsim.png
authors: Jiaqiang Zhu*, Lulu Shang*, Xiang Zhou
year: 2023
shortref: Zhu et al. (2023) 
journal: Genome Biology
type: statistical
doi: https://doi.org/10.1186/s13059-023-02879-z
---

# Abstract

**Overview**
Spatially resolved transcriptomics (SRT)-specific computational methods are often developed, tested, validated, and evaluated in silico using simulated data. Unfortunately, existing simulated SRT data are often poorly documented, hard to reproduce, or unrealistic. Single-cell simulators are not directly applicable for SRT simulation as they cannot incorporate spatial information. 

**Materials and Methods**
SRTsim, an SRT-specific simulator for scalable, reproducible, and realistic SRT simulations. SRTsim not only maintains various expression characteristics of SRT data but also preserves spatial patterns. 

**Results**
We illustrate the benefits of SRTsim in benchmarking methods for spatial clustering, spatial expression pattern detection, and cell-cell communication identification.

**Conclusion**
In conclusion, SRTsim is an effective simulation tool for SRT studies. SRTsim provides scalable, reproducible, and realistic simulations, while preserving expression characteristics and spatial patterns. We demonstrated the utility of SRTsim in evaluating the performance of spatial clustering methods, spatial expression analysis methods, and cell-cell communication identification methods.

<br />

<div class="middle">
    <img src="/assets/images/papers/SRTsim.png" alt="photo" width="500"/>
</div>

<br />

**Paper:**
<br />
[https://genomebiology.biomedcentral.com/articles/10.1186/s13059-023-02879-z](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-023-02879-z)

<br />

**Tutorial of SRTsim**
<br />
[SRTsim Tutorial](https://xzhoulab.github.io/SRTsim/)


