---
layout: paper
title: "Accurate imputation of pathway-specific gene expression in spatial transcriptomics with PASTA"
image: /assets/images/papers/PASTA.png
authors: Ruoxing Li, Peng Yang, Mauro Di Pilato, Jianjun Zhang, Christopher R. Flowers, Lulu Shang & Ziyi Li
year: 2025
shortref: Shang et al. (2025) 
journal: Nature Communications
type: method
doi: 
---

An essential task in spatial transcriptomics is identifying spatially variable genes (SVGs). Here, we present Celina, a statistical method for systematically detecting cell type-specific SVGs (ct-SVGs)—a subset of SVGs exhibiting distinct spatial expression patterns within specific cell types. Celina utilizes a spatially varying coefficient model to accurately capture each gene’s spatial expression pattern in relation to the distribution of cell types across tissue locations, ensuring effective type I error control and high power. Celina proves powerful compared to existing methods in single-cell resolution spatial transcriptomics and stands as the only effective solution for spot-resolution spatial transcriptomics. Applied to five real datasets, Celina uncovers ct-SVGs associated with tumor progression and patient survival in lung cancer, identifies metagenes with unique spatial patterns linked to cell proliferation and immune response in kidney cancer, and detects genes preferentially expressed near amyloid-β plaques in an Alzheimer’s model.


<br />

<div class="middle">
    <img src="/assets/images/papers/PASTA.png" alt="photo" width="500"/>
</div>

<br />

**Paper:**
<br />
Li, R., Yang, P., Di Pilato, M. et al. Accurate imputation of pathway-specific gene expression in spatial transcriptomics with PASTA. Nat Commun (2025). [https://www.nature.com/articles/s41467-025-67421-0](https://www.nature.com/articles/s41467-025-67421-0)


<br />